src/main/java/org/gridsuite/study/server/service/RemoteServices.java

```diff
@@ -0,0 +1,141 @@
+
+import com.fasterxml.jackson.core.JacksonException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.NullNode;
+import lombok.NonNull;
+import org.gridsuite.study.server.RemoteServicesProperties;
+import org.gridsuite.study.server.dto.ServiceStatusInfos;
+import org.gridsuite.study.server.dto.ServiceStatusInfos.ServiceStatus;
+import org.gridsuite.study.server.service.client.RemoteServiceName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.actuate.info.InfoEndpoint;
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * @author David Braquart <david.braquart at rte-france.com>
+ */
+@Service
+public class RemoteServices {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RemoteServices.class);
+
+    private static final String ACTUATOR_HEALTH_STATUS_JSON_FIELD = "status";
+    static final long REQUEST_TIMEOUT_IN_MS = 2000L;
+
+    private final ObjectMapper objectMapper;
+    private final RestTemplate restTemplate;
+    private final RemoteServicesProperties remoteServicesProperties;
+    private final InfoEndpoint infoEndpoint;
+    private final RemoteServices asyncSelf; //we need to use spring proxy-based bean
+
+    @Autowired
+    public RemoteServices(ObjectMapper objectMapper,
+                          RemoteServicesProperties remoteServicesProperties,
+                          @Lazy RemoteServices asyncRemoteServices,
+                          RestTemplateBuilder restTemplateBuilder,
+                          InfoEndpoint infoEndpoint) {
+        this.objectMapper = objectMapper;
+        this.remoteServicesProperties = remoteServicesProperties;
+        this.restTemplate = restTemplateBuilder
+                .setConnectTimeout(Duration.ofMillis(REQUEST_TIMEOUT_IN_MS))
+                .setReadTimeout(Duration.ofMillis(REQUEST_TIMEOUT_IN_MS))
+                .build();
+        this.infoEndpoint = infoEndpoint;
+        this.asyncSelf = asyncRemoteServices;
+    }
+
+    public List<ServiceStatusInfos> getOptionalServices() {
+        // parallel health status check for all services marked as "optional: true" in application.yaml
+        List<CompletableFuture<ServiceStatusInfos>> results = remoteServicesProperties.getServices().stream()
```

Comment:
Why don't we use `parallelStream` instead ?

Reply:
because we can't manage the thread pool used by streams (pool size, etc...)

Reply:
Where is it done here ? I don't see it. We don't use StudyServerExecutionService anyway right ?
Btw I think we can, even if it's not recommend : https://www.baeldung.com/java-when-to-use-parallel-stream#3-custom-thread-pool